diff --git a/src/snapshotpreview.cpp b/src/snapshotpreview.cpp
index e660489..9023211 100644
--- a/src/snapshotpreview.cpp
+++ b/src/snapshotpreview.cpp
@@ -85,6 +85,7 @@ void SnapshotPreview::init()
     m_progressDialog = 0;
     m_screenSaved = false;
     m_pixmapSet = false;
+    m_clipboard = QApplication::clipboard();
 
     ui = new Ui::SnapshotPreview;
     ui->setupUi(this);
@@ -102,6 +103,7 @@ void SnapshotPreview::init()
     ui->formatTextBtn->setIcon(iconLoader->loadIcon("draw-text", KIconLoader::Small));
 
     ui->cancelBtn->setIcon(iconLoader->loadIcon("dialog-cancel", KIconLoader::Small));
+    ui->copyBtn->setIcon(iconLoader->loadIcon("edit-copy", KIconLoader::Small));
     ui->saveBtn->setIcon(iconLoader->loadIcon("document-save-as", KIconLoader::Small));
     ui->uploadBtn->setIcon(KIcon("upload"));
 
@@ -124,6 +126,8 @@ void SnapshotPreview::init()
 
     connect(ui->cancelBtn, SIGNAL(clicked()),
             this, SLOT(close()));
+    connect(ui->copyBtn, SIGNAL(clicked()),
+            this, SLOT(slotCopy()));
     connect(ui->saveBtn, SIGNAL(clicked()),
             this, SLOT(slotSaveAs()));
     connect(ui->uploadBtn, SIGNAL(clicked()),
@@ -302,6 +306,13 @@ void SnapshotPreview::slotUpload()
     }
 }
 
+void SnapshotPreview::slotCopy()
+{
+    QPixmap p = getFinalPixmap();
+    m_clipboard->setPixmap(p);
+    m_screenSaved = true;
+}
+
 void SnapshotPreview::slotPrintUploadInfo(KJob *job, const QString &plain)
 {
     Q_UNUSED(job)
diff --git a/src/snapshotpreview.h b/src/snapshotpreview.h
index 671f616..e4d8b17 100644
--- a/src/snapshotpreview.h
+++ b/src/snapshotpreview.h
@@ -5,6 +5,7 @@
 #include <QPixmap>
 #include <QFrame>
 #include <QPointer>
+#include <QClipboard>
 
 namespace Ui {
     class SnapshotPreview;
@@ -52,6 +53,7 @@ private Q_SLOTS:
     void slotUploadResult(KJob *job);
     void slotPrintUploadInfo(KJob *job, const QString &plain);
     void slotCancelUpload();
+    void slotCopy();
     void slotSaveAs();
 
 private:
@@ -77,6 +79,7 @@ private Q_SLOTS:
     QPixmap m_lastGeneratedPixmap;
     bool m_screenSaved;
     bool m_pixmapSet;
+    QClipboard *m_clipboard;
 };
 
 #endif  //  SNAPSHOTPREVIEW_H
diff --git a/src/snapshotpreview.ui b/src/snapshotpreview.ui
index 046bfb0..b59108e 100644
--- a/src/snapshotpreview.ui
+++ b/src/snapshotpreview.ui
@@ -265,6 +265,13 @@
          </spacer>
         </item>
         <item>
+         <widget class="QPushButton" name="copyBtn">
+          <property name="text">
+           <string>Copy</string>
+          </property>
+         </widget>
+        </item>
+        <item>
          <widget class="QPushButton" name="uploadBtn">
           <property name="text">
            <string>Upload</string>
