diff --git a/vmmon-only/linux/hostif.c b/vmmon-only/linux/hostif.c
index 72e90aa..b33b1a7 100644
--- a/vmmon-only/linux/hostif.c
+++ b/vmmon-only/linux/hostif.c
@@ -2343,7 +2343,11 @@ isVAReadable(VA r)  // IN:
    int ret;
    
    old_fs = get_fs();
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 1, 0)
+   set_fs(KERNEL_DS);
+#else
    set_fs(get_ds());
+#endif
    r = APICR_TO_ADDR(r, APICR_VERSION);
    ret = HostIF_CopyFromUser(&dummy, (void*)r, sizeof(dummy));
    set_fs(old_fs);
@@ -2544,7 +2548,11 @@ HostIF_SemaphoreWait(VMDriver *vm,   // IN:
    }
 
    old_fs = get_fs();
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 1, 0)
+   set_fs(KERNEL_DS);
+#else
    set_fs(get_ds());
+#endif
 
    {
       struct poll_wqueues table;
@@ -2673,7 +2681,11 @@ HostIF_SemaphoreSignal(uint64 *args)  // IN:
    }
 
    old_fs = get_fs();
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 1, 0)
+   set_fs(KERNEL_DS);
+#else
    set_fs(get_ds());
+#endif
 
    /*
     * Always write sizeof(uint64) bytes. This works fine for eventfd and
