diff --git a/vmmon-only/linux/driver.c b/vmmon-only/linux/driver.c
index d79bcb4..015e4fc 100644
--- a/vmmon-only/linux/driver.c
+++ b/vmmon-only/linux/driver.c
@@ -117,7 +117,11 @@ static struct page *LinuxDriverNoPage(struct vm_area_struct *vma,
 #endif
 static int LinuxDriverMmap(struct file *filp, struct vm_area_struct *vma);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0)
+static void LinuxDriverPollTimeout(struct timer_list *clientData);
+#else
 static void LinuxDriverPollTimeout(unsigned long clientData);
+#endif
 
 static struct vm_operations_struct vmuser_mops = {
 #ifdef VMW_NOPAGE_2624
@@ -244,7 +248,11 @@ unregister_ioctl32_handlers(void)
  */
 
 static void
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0)
+LinuxDriverComputeTSCFreq(struct timer_list *data)  // IN:
+#else
 LinuxDriverComputeTSCFreq(unsigned long data)  // IN:
+#endif
 {
    Vmx86_GetkHzEstimate(&linuxState.startTime);
 }
@@ -287,9 +295,13 @@ init_module(void)
     */
 
    init_waitqueue_head(&linuxState.pollQueue);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0)
+   timer_setup(&linuxState.pollTimer, LinuxDriverPollTimeout, 0);
+#else
    init_timer(&linuxState.pollTimer);
    linuxState.pollTimer.data = 0;
    linuxState.pollTimer.function = LinuxDriverPollTimeout;
+#endif
 
    linuxState.fastClockThread = NULL;
    linuxState.fastClockFile = NULL;
@@ -365,9 +377,13 @@ init_module(void)
     */
 
    Vmx86_ReadTSCAndUptime(&linuxState.startTime);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0)
+   timer_setup(&tscTimer, LinuxDriverComputeTSCFreq, 0UL);
+#else
    init_timer(&tscTimer);
    tscTimer.data = 0;
    tscTimer.function = LinuxDriverComputeTSCFreq;
+#endif
    tscTimer.expires = jiffies + 4 * HZ;
    add_timer(&tscTimer);
 
@@ -903,7 +919,11 @@ LinuxDriverPoll(struct file *filp,  // IN:
  */
 
 static void
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0)
+LinuxDriverPollTimeout(struct timer_list *clientData)  // IN:
+#else
 LinuxDriverPollTimeout(unsigned long clientData)  // IN:
+#endif
 {
    LinuxDriverWakeUp(FALSE);
 }
diff --git a/vmmon-only/linux/hostif.c b/vmmon-only/linux/hostif.c
index 01d6a05..019a549 100644
--- a/vmmon-only/linux/hostif.c
+++ b/vmmon-only/linux/hostif.c
@@ -1825,7 +1825,11 @@ HostIFReadUptimeWork(unsigned long *j)  // OUT: current jiffies
  */
 
 static void
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0)
+HostIFUptimeResyncMono(struct timer_list *data)  // IN: ignored
+#else
 HostIFUptimeResyncMono(unsigned long data)  // IN: ignored
+#endif
 {
    unsigned long jifs;
    uintptr_t flags;
@@ -1887,8 +1891,12 @@ HostIF_InitUptime(void)
                   -(tv.tv_usec * (UPTIME_FREQ / 1000000) + 
                     tv.tv_sec * UPTIME_FREQ));
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0)
+   timer_setup(&uptimeState.timer, HostIFUptimeResyncMono, 0);
+#else
    init_timer(&uptimeState.timer);
    uptimeState.timer.function = HostIFUptimeResyncMono;
+#endif
    mod_timer(&uptimeState.timer, jiffies + HZ);
 }
 
