--- vmmon-only/linux/hostif.c.orig	2018-01-09 16:22:11.540376348 +0200
+++ vmmon-only/linux/hostif.c	2018-01-09 16:26:25.427618440 +0200
@@ -1642,9 +1642,18 @@
       BYTES_2_PAGES(vm->memInfo.hugePageBytes);
    unsigned int lockedPages = global_page_state(NR_PAGETABLE) +
                               global_page_state(NR_SLAB_UNRECLAIMABLE) +
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 8, 0)
+                              global_node_page_state(NR_UNEVICTABLE) +
+#else
                               global_page_state(NR_UNEVICTABLE) +
+#endif
                               hugePages + reservedPages;
-   unsigned int anonPages = global_page_state(NR_ANON_PAGES); 
+   unsigned int anonPages =
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 8, 0)
+      global_node_page_state(NR_ANON_PAGES); 
+#else
+      global_page_state(NR_ANON_PAGES); 
+#endif
    unsigned int swapPages = BYTES_2_PAGES(linuxState.swapSize);
 
    if (anonPages > swapPages) {
