diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 75f7b50..42d64c4 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -169,12 +169,6 @@ configure_file(${CMAKE_SOURCE_DIR}/data/desktopEntry/package/org.flameshot.Flame
 configure_file(${CMAKE_SOURCE_DIR}/data/appdata/org.flameshot.Flameshot.metainfo.xml
                ${CMAKE_CURRENT_BINARY_DIR}/share/metainfo/org.flameshot.Flameshot.metainfo.xml COPYONLY)
 
-configure_file(${CMAKE_SOURCE_DIR}/data/shell-completion/flameshot
-               ${CMAKE_CURRENT_BINARY_DIR}/share/bash-completion/completions/flameshot COPYONLY)
-
-configure_file(${CMAKE_SOURCE_DIR}/data/shell-completion/flameshot
-               ${CMAKE_CURRENT_BINARY_DIR}/share/zsh/site-functions/_flameshot COPYONLY)
-
 configure_file(${CMAKE_SOURCE_DIR}/data/dbus/org.flameshot.Flameshot.xml
                ${CMAKE_CURRENT_BINARY_DIR}/share/dbus-1/interfaces/org.flameshot.Flameshot.xml COPYONLY)
 
