diff --git a/CMakeLists.txt b/CMakeLists.txt
index a325f281..a77f118b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,7 +8,7 @@ set(CMAKE_CXX_FLAGS_DEBUG    "-O0")
 
 ################################################################################
 
-find_package(PythonLibs 3.3 REQUIRED)
+find_package(PythonLibs REQUIRED)
 
 if (WIN32)
     add_definitions("-DBOOST_PYTHON_STATIC_LIB")
@@ -16,20 +16,13 @@ endif()
 
 if (WIN32)
     find_package(Boost REQUIRED COMPONENTS python3)
-elseif (APPLE)
+else ()
-    foreach (PYTHON_NAME python3 python36 python37)
-        find_package(Boost QUIET COMPONENTS ${PYTHON_NAME})
+    foreach (PYTHON_NAME python python3.6 python3.7)
+        find_package(Boost REQUIRED COMPONENTS ${PYTHON_NAME})
         if (${Boost_FOUND})
             break()
         endif()
     endforeach()
-elseif (UNIX)
-    foreach (PYTHON_NAME python3 python-py36 python-py35 python-py34)
-        find_package(Boost QUIET COMPONENTS ${PYTHON_NAME})
-        if (${Boost_FOUND})
-            break()
-        endif()
-    endforeach()
 endif()
 
 if (NOT ${Boost_FOUND})
