diff --git a/CMakeScripts/DefineDependsandFlags.cmake b/CMakeScripts/DefineDependsandFlags.cmake
index 0e93631..214fb88 100644
--- a/CMakeScripts/DefineDependsandFlags.cmake
+++ b/CMakeScripts/DefineDependsandFlags.cmake
@@ -305,15 +305,6 @@ elseif(APPLE)
     list(APPEND INKSCAPE_CXX_FLAGS "-D_GNU_SOURCE")
 else()
     find_package(Boost 1.19.0 REQUIRED COMPONENTS filesystem)
-    # The package stacktrace_backtrace may not be available on all distros.
-    find_package(Boost 1.19.0 COMPONENTS stacktrace_backtrace)
-    if (BOOST_FOUND)
-        list(APPEND INKSCAPE_LIBS "-lbacktrace")
-        add_definitions("-DBOOST_STACKTRACE_USE_BACKTRACE")
-    else() # fall back to stacktrace_basic
-        find_package(Boost 1.19.0 REQUIRED COMPONENTS stacktrace_basic)
-        list(APPEND INKSCAPE_CXX_FLAGS "-D_GNU_SOURCE")
-    endif()
 endif()
 # enable explicit debug symbols
 set(CMAKE_ENABLE_EXPORTS ON)
